106-self-healing-feature-in-operator.md

```diff
@@ -0,0 +1,295 @@
+However, self-healing is currently disabled and disallowed in Strimzi.
+The `self-healing` properties were disabled since there was investigation missing around how self-healing would act if pods roll in middle of it/rebalance starts in or how the operator would know if self-healing is running or not.
+
+## Motivation
+
+Currently, any anomaly that the user is notified about would need to be fixed manually by using the `KafkaRebalance` custom resource.
+With smaller clusters, it is feasible to fix things manually. However, for larger ones it can be very time-consuming, or just not feasible, to fix all the anomalies on your own.
+It would be useful for users of Strimzi to be able to have these anomalies fixed automatically whenever they are detected.
+
+### Introduction to Self Healing
+
+![self-healing flow diagram](images/105-self-healing-flow.png)
+
+The above flow diagram depicts the self-healing process in Cruise Control.
+The anomaly detector manager detects an anomaly (using the detector classes) and forwards it to the notifier.
+The notifier classes like `SelfHealingNotifier`, `AlertaSelfHealingNotifier`, `SlackSelfHealingNotifier` etc. provides alerts to the users about the detected anomaly and also returns the action that needs to be taken on the anomaly i.e. whether to fix it, ignore it or delay it.
+
+#### Anomaly Detector Manager
+
+The anomaly detector manager helps in detecting the anomalies as well as handling them.
+It acts as a coordinator between the detector classes and the classes which will handle resolving the anomalies.
+Various detector classes like `GoalViolationDetector`, `DiskFailureDetector`, `KafkaBrokerFailureDetector` etc. are used for the anomaly detection, which runs periodically to check if the cluster has their corresponding anomalies or not.
+The frequency of this check can be changed via the `anomaly.detection.interval.ms` configuration.
+Detector classes have different mechanisms to detect their corresponding anomalies.
+For example, `KafkaBrokerFailureDetector` utilises Kafka Metadata API whereas `DiskFailureDetector` and `TopicAnomalyDetector` utilises Kafka Admin API.
+Furthermore, `MetricAnomalyDetector` and `GoalViolationDetector` use metrics to detect their anomalies.
+The detected anomalies can be of various types:
+* Goal Violation - This happens if certain optimization goals are violated (e.g. DiskUsageDistributionGoal etc.). These goals can be configured through the `self.healing.goals` configuration under the `spec.cruiseControl.config` section. These goals are independent of the manual rebalancing goals configured in the `KafkaRebalance` custom resource.
+* Topic Anomaly - Where one or more topics in cluster violates user-defined properties (e.g. some partitions are too large in disk).
+* Broker Failure - This happens when a non-empty broker crashes or leaves a cluster.
```

Comment:
"leaves a cluster" but for too long that's considered an error. The failure is not detected immediately everytime you do a scale down for example.